pkg/cloud-provider/cloudapi/azure/azure_security_test.go

```diff
@@ -0,0 +1,404 @@
+			mockazureService.EXPECT().virtualNetworks(gomock.Any()).Return(mockazureVirtualNetworksWrapper, nil).AnyTimes()
+			mockazureService.EXPECT().resourceGraph().Return(mockazureResourceGraph, nil)
+			mockazureVirtualNetworksWrapper.EXPECT().listAllComplete(gomock.Any()).AnyTimes()
+			mockazureResourceGraph.EXPECT().resources(gomock.Any(), gomock.Any()).Return(getResourceGraphResult(), nil).AnyTimes()
+			atAsg := network.ApplicationSecurityGroup{ID: &testATAsgID, Name: &atAsgID}
+			agAsg := network.ApplicationSecurityGroup{ID: &testAGAsgID, Name: &agAsgID}
+			asglist := []network.ApplicationSecurityGroup{agAsg, atAsg}
+
+			nsgrule := network.SecurityRule{
+				ID: &nsgID,
+				SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+					SourceApplicationSecurityGroups:      &[]network.ApplicationSecurityGroup{agAsg},
+					DestinationApplicationSecurityGroups: &[]network.ApplicationSecurityGroup{atAsg},
+					Priority:                             &testPriority,
+					SourcePortRange:                      &testSourcePortRange,
+					DestinationPortRange:                 &testDestinationPortRange,
+				},
+			}
+
+			nsg := network.SecurityGroup{
+				SecurityGroupPropertiesFormat: &network.SecurityGroupPropertiesFormat{
+					SecurityRules: &[]network.SecurityRule{nsgrule},
+				},
+				ID:   &testNsgID,
+				Name: &nsgID,
+			}
+
+			mockazureNsgWrapper.EXPECT().createOrUpdate(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nsg, nil).AnyTimes()
+			mockazureNsgWrapper.EXPECT().get(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nsg, nil).AnyTimes()
+			mockazureNwIntfWrapper.EXPECT().listAllComplete(gomock.Any()).AnyTimes()
+
+			mockazureAsgWrapper.EXPECT().createOrUpdate(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(agAsg, nil).AnyTimes()
+			mockazureAsgWrapper.EXPECT().get(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+			mockazureAsgWrapper.EXPECT().delete(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
+			mockazureAsgWrapper.EXPECT().listComplete(gomock.Any(), gomock.Any()).Return(asglist, nil).AnyTimes()
+
+			fakeClient = fake.NewClientBuilder().Build()
+			c = newAzureCloud(mockAzureServiceHelper)
+
+			vmSelector := []v1alpha1.VirtualMachineSelector{
+				{
+					VpcMatch: &v1alpha1.EntityMatch{MatchID: testVnetID01},
+					VMMatch:  []v1alpha1.EntityMatch{},
+				},
+			}
+
+			err := fakeClient.Create(context.Background(), secret)
+			Expect(err).Should(BeNil())
+			err = c.AddProviderAccount(fakeClient, account)
+			Expect(err).Should(BeNil())
+			selector.Spec.VMSelector = vmSelector
+			err = c.AddAccountResourceSelector(testAccountNamespacedName, selector)
+			Expect(err).Should(BeNil())
+
+			accCfg, _ := c.cloudCommon.GetCloudAccountByName(testAccountNamespacedName)
+			serviceConfig, _ := accCfg.GetServiceConfigByName(azureComputeServiceNameCompute)
+
+			vnetIDs := make(map[string]struct{})
```

Comment:
generally its not a good idea to combine everything in before test, unless its common for all. can we break this into smaller funcs/modules and explicitly call from the test. For example, define a function like setupAccount, addSelector etc.

Reply:
I think for asg and nsg part they can use for multiple times, it should be in the Return() part, and other create and addaccount part, it is always return an error in one line and then need a check error, I think this part also mostly the same steps in the aws tests